src/Microsoft.Diagnostics.DebugServices.Implementation/ServiceEvent.cs

```diff
@@ -0,0 +1,75 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Microsoft.Diagnostics.DebugServices.Implementation
+{
+    /// <summary>
+    /// The service event implementation
+    /// </summary>
+    public class ServiceEvent : IServiceEvent
+    {
+        private class EventNode : LinkedListNode
+        {
+            private readonly WeakReference<Action> _callback;
+
+            internal EventNode(Action callback)
+            {
+                _callback = new WeakReference<Action>(callback);
+            }
+
+            internal void Fire()
+            {
+                if (_callback.TryGetTarget(out Action target))
+                {
+                    target();
+                }
+                else
+                {
+                    Remove();
+                }
+            }
+        }
+
+        private class Unregister : IDisposable
+        {
+            private readonly EventNode _eventNode;
+            private readonly Action _callback;
+
+            internal Unregister(EventNode eventNode, Action callback)
+            {
+                _eventNode = eventNode;
+                _callback = callback;
+            }
+
+            void IDisposable.Dispose()
+            {
+                _eventNode.Remove();
+            }
+        }
+
+        private readonly LinkedListNode _events = new LinkedListNode();
+
+        public ServiceEvent()
+        {
+        }
+
+        public IDisposable Register(Action callback)
+        {
+            // Insert at the end of the list
+            var node = new EventNode(callback);
+            _events.InsertBefore(node);
+            return new Unregister(node, callback);
```

Comment:
This has become a very brittle contract. The registration token having the action means that if the user registers and event for updates but fails to hang the token for the duration as long as it needs to get updates, will never see updates because it will get evicted after it does out of scope and there's a GC.  The idea of making it an IDisposable is to make it explicit: please don't call me back once I've unregistered. A weakRef contract is brittle, and more so having an action that has the power to mutate the state of an object catured in the lambda. I still don't agree with the weak reference mechanism used.

Reply:
We should discuss some more.  I guess the eventing code I borrowed the weak reference idea/impl from also had a concept of a "strong" referenced event that wouldn't require the resulting registration token (IDisposable) to be rooted. I tried to simplify by only having the weak reference type of event that requires the registration token to be rooted.  You already found a bug (above in SOSHost.cs) where I forgot to do that.